ocaml/xapi/xapi_gpumon.ml

```diff
@@ -40,7 +42,34 @@ module Gpumon = Daemon_manager.Make (struct
     Xapi_systemctl.stop ~wait_until_success:false gpumon
 end)
 
-let with_gpumon_stopped = Gpumon.with_daemon_stopped
+let gpumon_m = Mutex.create ()
+
+let with_gpumon_stopped ?(timeout = 30.0) f =
+  match !Xapi_globs.nvidia_gpumon_detach with
+  | false ->
+      Gpumon.with_daemon_stopped ~timeout f
+  | true -> (
+      debug "%s: about to aquire lock" __FUNCTION__ ;
+      with_lock gpumon_m @@ fun () ->
+      let module GPU = Gpumon_client.Client.Nvidia in
+      let defer finally = Fun.protect ~finally in
+      let dbg = __FUNCTION__ in
+      match GPU.nvml_is_attached dbg with
+      | false ->
+          (* nothing to do, just execute f *)
+          debug "%s: NVML is detached; nothing to do" __FUNCTION__ ;
+          f ()
+      | true ->
+          (* detach, execute f, re-attach in any case. Be aware
+             that both xenopsd, xapi call /usr/lib/nvidia/sriov-manage,
+             which may stop stop gpumon *)
+          defer (fun () -> GPU.nvml_attach dbg) @@ fun () ->
+          debug "%s: about to detach NVML" __FUNCTION__ ;
+          GPU.nvml_detach dbg ;
```

Comment:
The defer is executed after the `detach` and `f`. All these functions are idempotent. However, I have wondered myself if it could happen that `f` does something that makes it impossible to re-attach NVML. The GPU is invisible for dom0 when it is in pass-trough mode.